### PR TITLE
Add XboxOne/DualShock layouts for Android

### PR DESCRIPTION
**Pull request by rytis-buz**
Added DualShock4GamepadAndroid and XboxOneGamepadAndroid layouts that derive from DualShockGamepad and XInputController respectively instead of deriving from AndroidGamepad
Fixed Right stick to use AXIS.Z and AXIS.RZ
Fixed triggers to always use Axis.Gas and Axis.Brake
Moved AndroidGamepadWithDpadButtons and AndroidGamepadWithDpadAxes from Json definition to class definition

### CODE REVIEW

File: Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidGameController.cs [hunk 1]
```diff
@@ -34,18 +42,26 @@ internal unsafe struct AndroidGameControllerState : IInputStateTypeInfo
...
+        [InputControl(name = "dpad/right", offset = 0, bit = 0, sizeInBits = 32, format = "FLT", parameters = "clamp=3,clampConstant=0,clampMin=0,clampMax=1", variants = kVariantDPadAxes)]
+        [InputControl(name = "dpad/left", offset = 0, bit = 0, sizeInBits = 32, format = "FLT", parameters = "clamp=3,clampConstant=0,clampMin=-1,clampMax=0,invert", variants = kVariantDPadAxes)]
+        [InputControl(name = "dpad/down", offset = ((uint)AndroidAxis.HatY - (uint)AndroidAxis.HatX) * sizeof(float), bit = 0, sizeInBits = 32, format = "FLT", parameters = "clamp=3,clampConstant=0,clampMin=0,clampMax=1", variants = kVariantDPadAxes)]
+        [InputControl(name = "dpad/up", offset = ((uint)AndroidAxis.HatY - (uint)AndroidAxis.HatX) * sizeof(float), bit = 0, sizeInBits = 32, format = "FLT", parameters = "clamp=3,clampConstant=0,clampMin=-1,clampMax=0,invert", variants = kVariantDPadAxes)]
+        [InputControl(name = "leftTrigger", offset = (uint)AndroidAxis.Brake * sizeof(float) + kAxisOffset, parameters = "clamp=1,clampMin=0,clampMax=1.0", variants = kVariantGamepad)]
```
Comment (rytis-buz) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidGameController.cs:50]: This clamping for triggers is questionable. There is a test that sets -1 and expects 0 as a result with a comment "Check if normalization works correctly". Clamping seemed to be the way to achieve that

**Comment by todi1856** [comment]
Could you add QA, since this is a sensitive change

**Comment by todi1856** [comment]
Missing entries in CHANGELOG

File: Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidGameController.cs [hunk 2]
```diff
@@ -34,18 +42,26 @@ internal unsafe struct AndroidGameControllerState : IInputStateTypeInfo
...
+        [InputControl(name = "rightStick", offset = (uint)AndroidAxis.Z * sizeof(float) + kAxisOffset, sizeInBits = ((uint)AndroidAxis.Rz - (uint)AndroidAxis.Z) * sizeof(float) * 8,
+            variants = kVariantGamepad)]
         [InputControl(name = "rightStick/x", variants = kVariantGamepad)]
         [InputControl(name = "rightStick/y", offset = ((uint)AndroidAxis.Rz - (uint)AndroidAxis.Z) * sizeof(float), variants = kVariantGamepad, parameters = "invert")]
-        [InputControl(name = "rightStick/up", offset = ((uint)AndroidAxis.Rz - (uint)AndroidAxis.Z) * sizeof(float), variants = kVariantGamepad, parameters = "invert,clamp=1,clampMin=-1.0,clampMax=0.0")]
-        [InputControl(name = "rightStick/down", offset = ((uint)AndroidAxis.Rz - (uint)AndroidAxis.Z) * sizeof(float), variants = kVariantGamepad, parameters = "invert=false,clamp=1,clampMin=0,clampMax=1.0")]
+        [InputControl(name = "rightStick/up", offset = ((uint)AndroidAxis.Rz - (uint)AndroidAxis.Z) * sizeof(float), variants = kVariantGamepad,
```
Comment (todi1856) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidGameController.cs]: no change here ? can you revert, same for  rightStick

File: Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidGameController.cs [hunk 3]
```diff
@@ -152,13 +168,13 @@ namespace UnityEngine.InputSystem.Android
     /// Other exotic gamepads have different mappings
     ///  Xbox Gamepad (for ex., Microsoft X-Box One pad (Firmware 2015)) mapping (Note mapping: L2/R2/Right Stick)
     ///  Left Stick -> AXIS_X(0) / AXIS_Y(1)
-    ///  Right Stick -> AXIS_RX (12) / AXIS_RY(13)
+    ///  Right Stick -> AXIS_Z (11) / AXIS_RZ(14)
```
Comment (todi1856) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidGameController.cs]: So it seems, this comment is no longer relevent, as mapping matches the comment scope above next to /// Most of the gamepads: 

if so, can you delete this scope, and add Xbox Controller in the list above (after Sony Interactive Entertainment Wireless), also could you please add device name and android os version next to the device just as an additional info

**Review by todi1856 — CHANGES_REQUESTED** [review]
Looks good, tested myself Dualschock on Google Pixel 2 XL with Android 11, works fine.

Please add QA, after QA approves, I'll approve too

**Review by todi1856 — CHANGES_REQUESTED** [review]
Please add entries to CHANGELOG

**Review by todi1856 — APPROVED** [review]

**Comment by renaldasrep** [comment]
Sorry it's taking a while, we're trying to get a hold of required controllers, should be able to have them at hand by the end of the week

**Review by renaldasrep — APPROVED** [review]
Tested on multiple android devices using DualShock 4 and the Xbox Series X controller
No issues found apart DualShock 4 working pretty randomly on Android 8, but it is documented now that it is not officially supported (notes added in this PR)

**Comment by rytis-buz** [comment]
@Rene-Damm The PR has been open for 2 months and now have conflicts. Can we move forward with it?

**Comment by ekcoh** [comment]
@rytis-buz @todi1856 @renaldasrep sorry for the late follow-up on this PR but we (@andrew-oc, @ekcoh)  have looked through it now and would be happy to merge it to develop. We did some minor modifications on our end and just want to check that yo uare onboard with this happening now before we merge.

**Comment by todi1856** [comment]
looks good

**Comment by rytis-buz** [comment]
Looks good to me

**Pull request merged**